### ISSUE TITLE
miniupnpd: Time for a new release?

### ISSUE THREAD

**Issue by pali** [miniupnp/miniupnp#429]
Last release of miniupnpd is from year 2018 and in git master branch there are lot changes, including fix for linux kernel 5.0. It would be nice to see a new release of miniupnpd so distributions do not have to build miniupnpd from git or trying to backport commits from git to older release tarball versions.

**Comment by miniupnp** [comment]
I wish I could have a better netfilter nftables backend for the 2.2 release

**Comment by Neustradamus** [comment]
I think yes!

But @miniupnp, have you seen https://github.com/miniupnp/miniupnp/pull/409?

**Comment by miniupnp** [comment]
yes, but I cannot merge it as it. Too much different changes mixed together.

I'm currently adding privilege dropping for linux via libcap.
I think I will provide a way to choose between libcap and libcap-ng.

**Comment by The-any-Key** [comment]
A new release would be good as most antivirus now treat this app as a unsafe application just because there is no new release of it. They treat this app as a "non actively maintained" app and therefore block it.

![image](https://user-images.githubusercontent.com/6436102/80855837-c8a2d200-8c44-11ea-8f58-f3afb6a79b56.png)

**Comment by miniupnp** [comment]
Well I don't really care with developers that are not able to use the library and embed the command line tool their application directory.

That is indeed a very unsafe way to do

**Comment by miniupnp** [comment]
miniupnpd-2.2.0-RC0.tar.gz released

**Comment by Neustradamus** [comment]
@miniupnp: Nice!
But please do not forget my ticket before stable release, it is really easy and take few minutes.

**Comment by paul-chambers** [comment]
@miniupnp  - is your concern with the issue raised in #422 ? I can take another look, though time is scarce again, having just started a new job.

**Comment by miniupnp** [comment]
> @miniupnp - is your concern with the issue raised in #422 ? I can take another look, though time is scarce again, having just started a new job. The changes required are likely to be significant.

if the change are significant, that may be for a next release

**Comment by paul-chambers** [comment]
In my opinion, to 'do it right' would mean shifting from the original 'iptables-like' approach of the original design to one based on nftables maps. That's not a small change, though the code itself and the 'footprint' in the tables is likely to be cleaner.

While the current implementation isn't how I'd have designed it, it does 'work', whereas it didn't in the prior release.

**Comment by emojifreak** [comment]
I applied clang sanitizers on miniupnpd on Linux-nftables platform, and found many errors as #459 . It is said that if you see one cockroach, there are a thousand more. The linux-nftables part of miniupnpd could have particularly worse quality than other parts, or miniupnpd on other platforms has similar number of potential errors. I have little idea of it.

On the other hand, miniupnpd has received several security advisories like CVE in recent past. It seems safer to apply sanitizers on other platforms, specifically [UBSan](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html), [LSan](https://clang.llvm.org/docs/LeakSanitizer.html), [ASan](https://clang.llvm.org/docs/AddressSanitizer.html) and [MSan](https://clang.llvm.org/docs/MemorySanitizer.html), **before making next official release**, to avoid future security advisories.

As [LSan](https://clang.llvm.org/docs/LeakSanitizer.html) and [UBSan](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html) have little penalty in execution, they could be included in the default compilation option on some platforms in official releases. 

[UBSan](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html), [LSan](https://clang.llvm.org/docs/LeakSanitizer.html) and  [ASan](https://clang.llvm.org/docs/AddressSanitizer.html) can be used with gcc in addition to LLVM clang.

**Comment by miniupnp** [comment]
It looks like memory sanitizer is yielding a lot of false positives

**Comment by JAMESMTL** [comment]
in miniupnpd 2.2.0-RC1 master-fca45a6 I am seeing an issue upon expiry of a rule when using nftables. the rule in the ip miniupnpd table is removed but not the rule in the inet miniupnpd table. log shows rule removed.

thanks

**Comment by emojifreak** [comment]
`nft list ruleset` shows the following after rule expiration and it looks the same as manual deletion of a rule by `upnpc -d`:

```
table inet miniupnpd {
        chain forward {
                type filter hook forward priority -25; policy accept;
        }
}
table ip miniupnpd {
        chain prerouting {
                type nat hook prerouting priority dstnat; policy accept;
        }

        chain postrouting {
                type nat hook postrouting priority srcnat; policy accept;
        }
}
table ip6 miniupnpd {
        chain prerouting {
                type nat hook prerouting priority dstnat; policy accept;
        }

        chain postrouting {
                type nat hook postrouting priority srcnat; policy accept;
        }
}
```

**Comment by miniupnp** [comment]
> in miniupnpd 2.2.0-RC1 master-fca45a6 I am seeing an issue upon expiry of a rule when using nftables. the rule in the ip miniupnpd table is removed but not the rule in the inet miniupnpd table. log shows rule removed.
> 
> thanks

please open a new issue in github if needed

**Comment by Neustradamus** [comment]
Of course before the new release, it will be very important to change it:
- https://github.com/miniupnp/miniupnp/issues/436

If anyone can prove the importance of the thing to Thomas, it will be excellent.
I really need help.
It is a quickly operation.

**Comment by risicle** [comment]
A simple point release (2.1.1?) with just the CVE fixes would be a lot better than nothing.

**Comment by miniupnp** [comment]
fixed by 56c66b5472d68cb6bd6903ee0a32f420f4f542c4
